service/history/service.go

```diff
@@ -138,37 +137,19 @@ func (s *Service) Stop() {
 		return
 	}
 
-	// initiate graceful shutdown :
-	// 1. remove self from the membership ring
-	// 2. wait for other members to discover we are going down
-	// 3. stop acquiring new shards (periodically or based on other membership changes)
-	// 4. wait for shard ownership to transfer (and inflight requests to drain) while still accepting new requests
-	// 5. Reject all requests arriving at rpc handler to avoid taking on more work except for RespondXXXCompleted and
-	//    RecordXXStarted APIs - for these APIs, most of the work is already one and rejecting at last stage is
-	//    probably not that desirable. If the shard is closed, these requests will fail anyways.
-	// 6. wait for grace period
-	// 7. force stop the whole world and return
-
-	const gossipPropagationDelay = 400 * time.Millisecond
-	const shardOwnershipTransferDelay = 5 * time.Second
-	const gracePeriod = 2 * time.Second
-
-	remainingTime := s.config.ShutdownDrainDuration()
-
 	logger.Info("ShutdownHandler: Evicting self from membership ring")
 	_ = s.membershipMonitor.EvictSelf()
-	s.healthServer.SetServingStatus(serviceName, healthpb.HealthCheckResponse_NOT_SERVING)
 
-	logger.Info("ShutdownHandler: Waiting for others to discover I am unhealthy")
-	remainingTime = s.sleep(gossipPropagationDelay, remainingTime)
+	if delay := s.config.ShutdownDrainDuration(); delay > 0 {
+		s.logger.Info("ShutdownHandler: delaying for shutdown drain",
+			tag.NewDurationTag("shutdownDrainDuration", delay))
+		time.Sleep(s.config.ShutdownDrainDuration())
+	}
+
+	s.healthServer.SetServingStatus(serviceName, healthpb.HealthCheckResponse_NOT_SERVING)
```

Comment:
What effect does moving this have?

Reply:
It keeps external health checkers (like kubernetes readiness probes using https://github.com/grpc-ecosystem/grpc-health-probe ) from thinking that the instances has gone unhealthy while we're intentionally pausing, and prematurely terminating the instance.